### PR TITLE
WebGLRenderer: Change scissor, viewport functions to use "round" instead of "floor"

### LINKED ISSUE #27655: WebGLRenderer: Setting Scissor and Viewport is imprecise due to rounding issues
### Description

For [three-gpu-pathtracer](https://github.com/gkjohnson/three-gpu-pathtracer) I've been rendering the image in tiles for the sake of performance but found that it's not possible to set the exact pixels that the renderer rasterize due to the multiplication with the pixel ration and subsequent [rounding that occurs](https://github.com/mrdoob/three.js/blob/1b35d645db0070803e48fb5ba302478c99ac99d7/src/renderers/WebGLRenderer.js#L516) in the setScissor and setViewport functions.

This means that when rendering tiles side by side the rounding can result in pixels being missed when rendering tiles. For three-gpu-pathtracer I've switched to using the WebGLRenderTarget's .scissor property since it affords setting the sciossor via exact pixel values.

### Reproduction steps

1. Try to render in a tiled fashion with a combination of pixel ratio and dimensions that do not cleanly divide
2. See that the tiles miss rows and columns of pixels due to rounding error

### Code

See live example.

### Live example

Demo with selected pixel ratio and dimensions that result in the skipped pixels.

https://jsfiddle.net/cfhwous9/

### Screenshots

<div>image</div>

<div>image</div>

https://twitter.com/mrdoob/status/1752366500621746617

### Version

r160

### Device

Desktop, Mobile

### Browser

Chrome, Firefox, Safari, Edge

### OS

Windows, MacOS, Android

### PR DESCRIPTION
**Pull request by gkjohnson**
Fixed #27655

**Description**

Use "round" instead of "floor" to compute the scissor and viewport functions to prevent rounding errors. From the WebGL spec there doesn't seem to be any requirement that the max value be within the width of the canvas so no value clamp has been added.

### CODE REVIEW

**Comment by mrdoob** [comment]
What I started doing for these cases is to download the screenshot that the CI generated and add it to the PR... 😇

You can get it here:
https://github.com/mrdoob/three.js/actions/runs/7814629999/artifacts/1227023113

**Comment by WestLangley** [comment]
https://github.com/mrdoob/three.js/pull/18326 may need to be revisited if this is merged. It was required because of the use of '.floor()'

**Comment by WestLangley** [comment]
https://github.com/mrdoob/three.js/pull/18326 may need to be revisited if this is merged. It was required because of the use of `.floor()`.

**Comment by gkjohnson** [comment]
> https://github.com/mrdoob/three.js/pull/18326 may need to be revisited if this is merged. It was required because of the use of .floor().

This isn't something I'd be able to test, unfortunately. cc @elalish

**Comment by elalish** [comment]
Well, it should be pretty obvious with a Pixel and this example: https://github.com/mrdoob/three.js/issues/18265#issuecomment-571085967

I can't test just now either, but I will once I pull this into MV.

**Comment by gkjohnson** [comment]
Looking at the PR - it might be best to set the viewport settings on the render target rather than the renderer. Setting the viewport on the render target itself allows you to set exact pixel values instead of dealing with these rounding issues and DPR.

The discrepancy in behavior when dealing with WebGLRenderTarget.viewport and WebGLRenderer.setViewport (and scissor) is fairly confusing and it might be worth looking into how to simplify this another time.

**Comment by donmccurdy** [comment]
> From the WebGL spec there doesn't seem to be any requirement that the max value be within the width of the canvas so no value clamp has been added.

Examples like https://threejs.org/examples/#webgl_multiple_elements rely on being able to set scissor and viewport values that extend partially offscreen, so +1 for not clamping this. 🙂

**Comment by Mugen87** [comment]
@gkjohnson Do you mind temporarily updating the build files in this PR? I can make a test with a Pixel 4a which has a device pixel ratio of `2.75`. If there is an issue with PMREM, examples like [webgl_materials_envmaps_exr](https://threejs.org/examples/webgl_materials_envmaps_exr) or [webgl_materials_physical_clearcoat](https://threejs.org/examples/webgl_materials_physical_clearcoat) should show the banding artifacts from #18265.

**Review by Mugen87 — APPROVED** [review]

File: src/renderers/WebGLRenderer.js [hunk 1]
```diff
@@ -489,7 +489,7 @@ class WebGLRenderer {
 
 			}
 
-			state.viewport( _currentViewport.copy( _viewport ).multiplyScalar( _pixelRatio ).floor() );
+			state.viewport( _currentViewport.copy( _viewport ).multiplyScalar( _pixelRatio ).round() );
```
Comment (WestLangley) [inline src/renderers/WebGLRenderer.js:492]: In at least six other instances in the code, `floor()` is still used when applying `pixelRatio`. Was this an oversight, or was this intentional?

/ping @Mugen87
/ping @gkjohnson

Comment (Mugen87) [reply src/renderers/WebGLRenderer.js:492]: To me that is an oversight. When looking at the code, the usage in `setRenderTarget()` should be updated to `round()` as well.

https://github.com/mrdoob/three.js/blob/0bc990f5f728552c6cc3931d1b44afa9718aef40/src/renderers/WebGLRenderer.js#L2286-L2287

Not sure about `getDrawingBufferSize()` though.

https://github.com/mrdoob/three.js/blob/0bc990f5f728552c6cc3931d1b44afa9718aef40/src/renderers/WebGLRenderer.js#L412

Depending on how we decide, the code in `examples/jsm/renderers/common/Renderer.js` should be updated as well.

Comment (WestLangley) [reply src/renderers/WebGLRenderer.js:492]: /bump @Mugen87

Comment (WestLangley) [reply src/renderers/WebGLRenderer.js:492]: >From the WebGL spec there doesn't seem to be any requirement that the max value be within the width of the canvas so no value clamp has been added.

In WebGPU, attempting to set a scissor rectangle outside the bounds of the render target results in a validation error.

Comment (Mugen87) [reply src/renderers/WebGLRenderer.js:492]: Feel free to file a PR with your suggested change 👍 .

**Pull request merged**